### ISSUE TITLE
Proposal for build system under many system configuration testing

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#15350]
We support a large number of build system configurations, but only a subset of them are actually tested in CI. This means that patches that are build fixes for exotic build configurations are difficult to check for correctness. Often, a fix for one build configuration, will break another build configuration. Recent example: #11563 update for AppleClang broke PPC builds (and it also broke OpenMP support entirely, although that's another matter.) However, we can't just keep adding build configurations for every system we might want to support, since that will be ruinously expensive to CI.

However, I have made the following observations:

* Many useful-to-catch build problems happen during cmake stage. So we don't actually need to build the entire system, we can just cmake, and see if cmake succeeded or not (and found the libraries we expect it to find.)
* You don't really need to build all of PyTorch to see if an external library is working or not; all you need is a small test program that uses the external library.

So, we should be able to make a much faster build which we can use solely to smoketest build system problems.

cc @ezyang @seemethere @malfet @walterddr @pytorch/pytorch-dev-infra

**Comment by smessmer** [comment]
We could combine this with running a full test suite on them from time to time (e.g. for releases). That'd also be a check if this approach works. If the full tests break too often for things that haven't been caught in CI, we know we have to run more of them in CI.

**Comment by ngimel** [comment]
This is still an issue (probably more of an issue)

**Comment by zhouzhuojie** [comment]
We are thinking of supporting

1. "Validation sets" (i.e. modular set of tests) brought up by @mruberry
2. OSS CI tests dispatcher to dynamically match the PR context (labels/description/comments/diff) with the validation sets.

More detail to come. Hopefully, effort like this can start addressing the issue.

**Comment by ezyang** [comment]
@zhouzhuojie That sounds orthogonal to this issue, which is primarily about testing our cmake code.

**Comment by ZainRizvi** [comment]
Assuming this is already addressed. Please reopen if that's not the case
